### ISSUE TITLE
Replace `ConnectedPoint` with `Dialer` in `libp2p_swarm::DialError::WrongPeerId`

### ISSUE THREAD

**Issue by dmitry-markin** [libp2p/rust-libp2p#2767]
## Description

Change the [`endpoint`](https://github.com/libp2p/rust-libp2p/blob/62622a1bad70381e56237b4da4d9abcf2b1a0e24/swarm/src/lib.rs#L1444) type to `Dialer`, extracting it from [`ConnectedPoint`](https://github.com/libp2p/rust-libp2p/blob/62622a1bad70381e56237b4da4d9abcf2b1a0e24/core/src/connection.rs#L116).

## Motivation

Currently in case of [`DialError`](https://github.com/libp2p/rust-libp2p/blob/62622a1bad70381e56237b4da4d9abcf2b1a0e24/swarm/src/lib.rs#L1423) with `WrongPeerId` the [`ConnectedPoint`](https://github.com/libp2p/rust-libp2p/blob/62622a1bad70381e56237b4da4d9abcf2b1a0e24/core/src/connection.rs#L116) is returned, which can contain `Listener`, which can't happen, because we are the dialing side. Replacing `ConnectedPoint` with `Dialer` can help to resolve this ambiguity and eliminate client code assertions.

`ConnectedPoint` endpoint type in `WrongPeerId` was introduced in [this PR](https://github.com/libp2p/rust-libp2p/pull/2428#discussion_r786066319), which was merged [here](https://github.com/libp2p/rust-libp2p/pull/2441).

The original discussion that lead to this issue is in [the substrate PR](https://github.com/paritytech/substrate/pull/11872#discussion_r926094382) that changes the error printed when we connect to a bootnode that provides a different than expected peer id.

## Are you planning to do it yourself in a pull request?

~~Maybe.~~ Yes.

**Comment by mxinden** [comment]
That sounds reasonable to me. Thanks @dmitry-markin for the detailed report. 

I am not going to work on this any time soon. I am happy to help you implement it though. Given your "maybe", are you still interested?

**Comment by dmitry-markin** [comment]
Yes, I think I can look into it in a week or so. The one thing I worry about is that these changes will definitely break the API, and I don't know how to handle it. I.e, I can update substrate, but we need to somehow notify other clients.

**Comment by mxinden** [comment]
Given that this is not a subtle change, but a change at the type system, I think users will (a) notice early and (b) can consult the changelog to get more advice. In other words, I am not worried about the breaking API change @dmitry-markin.

**Comment by dmitry-markin** [comment]
It seems the cause for `DialError::WrongPeerId` to contain `ConnectedPoint`, and not `Dialer` or just `Multiaddr` is that [`PendingOutboundConnectionError`](https://github.com/libp2p/rust-libp2p/blob/62622a1bad70381e56237b4da4d9abcf2b1a0e24/swarm/src/connection/error.rs#L89) is based on [`PendingConnectionError`](https://github.com/libp2p/rust-libp2p/blob/62622a1bad70381e56237b4da4d9abcf2b1a0e24/swarm/src/connection/error.rs#L98) also used for [`PendingInboundConnectionError`](https://github.com/libp2p/rust-libp2p/blob/62622a1bad70381e56237b4da4d9abcf2b1a0e24/swarm/src/connection/error.rs#L93). So they both share `ConnectedPoint` internally, even though `ConnectedPoint::Listener` is not relevant for `PendingOutboundConnectionError`/`DialError`.

To fix the API issue we can drop the `Listener` branch when converting `PendingOutboundConnectionError` into `DialError` [here](https://github.com/libp2p/rust-libp2p/blob/62622a1bad70381e56237b4da4d9abcf2b1a0e24/swarm/src/lib.rs#L1457).

~~I see two ways of doing this. First is to extract `ConnectedDialer`, `PendingDialer`, and `Listener` (listeners are the same in both enums) from `ConnectedPoint` & `PendingPoint`. Then replace `ConnectedPoint` with `ConnectedDialer` in `DialError`. This requires a lot of modifications throughout the codebase and significantly changes the API.~~ EDIT: I don't think this makes sense :-)

Another option is to keep `ConnectedPoint` and `PendingPoint` enums the same, and just modify `DialError` to include something instead `ConnectedPoint`. This can be either independent new `ConnectedDialer` struct, or even a plain `Multiaddr`.

I would go with the second option, but I don't understand whether we need to report the [`role_override`](https://github.com/libp2p/rust-libp2p/blob/62622a1bad70381e56237b4da4d9abcf2b1a0e24/core/src/connection.rs#L141) when reporting the `DialError` (i.e., use the `ConnectedDialer` struct), or reporting a `Multiaddr` will be enough.

@mxinden what do you think?

**Comment by dmitry-markin** [comment]
I've implemented the simplest solution in https://github.com/libp2p/rust-libp2p/pull/2793.

**Comment by mxinden** [comment]
> It seems the cause for `DialError::WrongPeerId` to contain `ConnectedPoint`, and not `Dialer` or just `Multiaddr` is that [`PendingOutboundConnectionError`](https://github.com/libp2p/rust-libp2p/blob/62622a1bad70381e56237b4da4d9abcf2b1a0e24/swarm/src/connection/error.rs#L89) is based on [`PendingConnectionError`](https://github.com/libp2p/rust-libp2p/blob/62622a1bad70381e56237b4da4d9abcf2b1a0e24/swarm/src/connection/error.rs#L98) also used for [`PendingInboundConnectionError`](https://github.com/libp2p/rust-libp2p/blob/62622a1bad70381e56237b4da4d9abcf2b1a0e24/swarm/src/connection/error.rs#L93). So they both share `ConnectedPoint` internally, even though `ConnectedPoint::Listener` is not relevant for `PendingOutboundConnectionError`/`DialError`.

I would be in favor of doing as much as possible at compile time. Would splitting `PendingConnectionError` be an option?

**Comment by dmitry-markin** [comment]
> I would be in favor of doing as much as possible at compile time. Would splitting PendingConnectionError be an option?

Yes, this seems reasonable, I'll check it.

**Comment by akaladarshi** [comment]
Hey @jxs 

Can I pick this issue?

**Comment by jxs** [comment]
Hi @akaladarshi yes! that would be great, Thanks! 🚀
